File: sdk/python/src/graft/modules/stepper.py
"""GRAFT Stepper Module"""

from __future__ import annotations

import struct
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from graft.device import GraftDevice

from graft.core.constants import CMD


class GraftStepper:
    def __init__(self, device: GraftDevice, module_id: int = 0) -> None:
        self._device = device
        self._mod_id = module_id

    async def config(self, step_pin: int, dir_pin: int, en_pin: int, steps_per_rev: int = 200) -> None:
        payload = bytes([self._mod_id, 0x01, step_pin, dir_pin, en_pin,
                         steps_per_rev & 0xFF, (steps_per_rev >> 8) & 0xFF])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def move(self, steps: int, speed_hz: int) -> None:
        payload = bytes([self._mod_id, 0x02]) + struct.pack("<ih", steps, speed_hz)
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def move_to(self, position: int, speed_hz: int) -> None:
        payload = bytes([self._mod_id, 0x03]) + struct.pack("<ih", position, speed_hz)
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def stop(self) -> None:
        payload = bytes([self._mod_id, 0x04])
        await self._device._send_command(CMD.MOD_CMD, payload)

File: sdk/js/src/modules/stepper.ts
/**
 * GRAFT Stepper Module — Typed JS wrapper
 */

import type { GraftDevice } from '../device.js'

export class GraftStepper {
  private _device: ReturnType<GraftDevice['module']>

  constructor(device: GraftDevice) {
    this._device = device.module('stepper')
  }

  /** Configure stepper pins and steps per revolution. */
  async config(stepPin: number, dirPin: number, enPin: number, stepsPerRev = 200): Promise<void> {
    const buf = new Uint8Array(5)
    buf[0] = stepPin; buf[1] = dirPin; buf[2] = enPin
    buf[3] = stepsPerRev & 0xFF; buf[4] = (stepsPerRev >> 8) & 0xFF
    await this._device.cmd(0x01, buf)
  }

  /** Move relative number of steps at given speed (Hz). */
  async move(steps: number, speedHz: number): Promise<void> {
    const buf = new Uint8Array(6)
    const view = new DataView(buf.buffer)
    view.setInt32(0, steps, true)
    view.setUint16(4, speedHz, true)
    await this._device.cmd(0x02, buf)
  }

  /** Move to absolute position at given speed (Hz). */
  async moveTo(position: number, speedHz: number): Promise<void> {
    const buf = new Uint8Array(6)
    const view = new DataView(buf.buffer)
    view.setInt32(0, position, true)
    view.setUint16(4, speedHz, true)
    await this._device.cmd(0x03, buf)
  }

  /** Emergency stop. */
  async stop(): Promise<void> {
    await this._device.cmd(0x04)
  }

  /** Listen for movement completion events. */
  onDone(handler: (position: number) => void): void {
    this._device.onEvent((code, data) => {
      if (code === 0x01 && data.length >= 4) {
        const view = new DataView(data.buffer, data.byteOffset, data.byteLength)
        handler(view.getInt32(0, true))
      }
    })
  }
}
